hi-ml-multimodal/src/health_multimodal/image/model/transformer.py

```diff
@@ -0,0 +1,262 @@
+        trunc_normal_(self.cls_token, std=.02)
+        trunc_normal_(self.type_embed, std=.02)
+
+        # Positional embeddings 1 x L x C (L: Sequence length, C: Feature dimension)
+        self.pos_drop = nn.Dropout(p=0.10)
+        pos_embed_class = SinePositionEmbedding(embedding_dim=input_dim // 2, normalize=True)
+        pos_embed = pos_embed_class(mask=torch.ones([1, grid_shape[0], grid_shape[1]]))  # 1 x L x C
+        self.register_buffer("pos_embed", pos_embed, persistent=False)
+
+        # Initialisation
+        self.apply(self._init_weights)
+
+    def no_weight_decay(self) -> Set[str]:
+        return {'type_embed', 'cls_token'}
+
+    def forward(self, current_image: torch.Tensor, previous_image: Optional[torch.Tensor] = None) -> torch.Tensor:
+        B, C, H, W = current_image.shape
+        assert H == self.grid_shape[0] and W == self.grid_shape[1], "Input and grid shapes do not match"
+
+        # Flatten patch embeddings
+        if previous_image is not None:
+            assert previous_image.shape == current_image.shape, "current_image and previous_image shapes do not match"
+            previous_image = previous_image.view(B, C, H * W).transpose(1, 2)
+        current_image = current_image.view(B, C, H * W).transpose(1, 2)
+        pos_embed = self.pos_embed.repeat(B, 1, 1)  # type: ignore
+
+        # Final token activations (B x 2L x C)
+        token_features = self.forward_after_reshape(x=current_image, pos_embed=pos_embed, x_previous=previous_image)
+
+        # Extract the patch features of current image
+        cur_img_token_id = 1 if self.use_cls_token else 0
+        current_token_features = token_features[:, cur_img_token_id:self.num_patches+cur_img_token_id]
+        current_patch_features = current_token_features.transpose(1, 2).view(B, C, H, W)
+
+        return current_patch_features
+
+    def forward_after_reshape(self,
+                              x: torch.Tensor,
+                              pos_embed: torch.Tensor,
+                              x_previous: Optional[torch.Tensor] = None) -> torch.Tensor:
+        B, L, _ = x.shape  # Batch, Sequence length, Feature dimension
+
+        # Positional and type embeddings
+        type_embed = self.type_embed[0].expand(B, L, -1)
+        if x_previous is not None:
+            x = torch.cat((x, x_previous), dim=1)
+            pos_embed = torch.cat((pos_embed, pos_embed), dim=1)
+            prev_type_embed = self.type_embed[1].expand(B, L, -1)
+            type_embed = torch.cat((type_embed, prev_type_embed), dim=1)
+
+        # Add positional and type embeddings (used in query and key matching)
+        pos_and_type_embed = pos_embed + type_embed
+
+        # Positional dropout and CLS token if required
+        x = self.pos_drop(x)
+        if self.use_cls_token:
+            cls_tokens = self.cls_token.expand(x.shape[0], -1, -1)
+            x = torch.cat((cls_tokens, x), dim=1)
```

Comment:
Remove? Currently not covered.

Reply:
Updated - We will need to update the checkpoint as well, but I think it was a good idea since the corresponding parameters were not trained.